### DISCUSSION TITLE
Adding a top-level `build-file` argument to `git_override`

### DISCUSSION THREAD

**Discussion by nicholasjng** [bazelbuild/bazel#21336] [Ideas]
I was wondering if it would make sense to include a build file slot in the `git_override` method (and possibly the others too) in MODULE.bazel.

My angle is trying to pull in a git dependency that does not have a Bazel integration, and thus needs a proper build file marking it as a package. This used to work by checking in a build file and giving it to the `git_repository` function, but bzlmod's overrides do not have such a slot.

I know that I can probably keep my build file from the workspace era in my main repo and make a patch out of it (does anyone have a nice shell snippet for that by the way, generating a whole-file patch of a checked-in file?), but I think this would be so much of a UX improvement that I'd like to see it. How do others feel about this?

**Comment by fmeum** [comment]
You can still use the `git_repository` repository rule from your module file via `use_repo_rule`. `git_override` is less convenient anyway as you will also have to specify a `MODULE.bazel` file for your dependency. With `use_repo_rule`, it just inherits the dependencies of your root module.

**Reply by nicholasjng** [reply]
Thanks! But then what is the point of `{git,archive}_override`? I thought those were replacements for their workspace counterparts in bzlmod.

I'm currently migrating a repo to bzlmod that has no Bazel integration, but only has BCR-available deps, so its MODULE.bazel file is (almost) trivial. Is there any downside to using `use_repo_rule` then over the "module-native" apis?

**Reply by fmeum** [reply]
A `git_override` (together with a `bazel_dep`), brings in a Bazel *module*, which is more than just a repo: such a module can also specify its own dependencies via `bazel_dep`. If you want just a repo (every module is also a repo), you should go with `git_repository`.

**Reply by nicholasjng** [reply]
I see. What's the heuristic for module vs. repo, then?

My current understanding is "If the dependency has Bazel functionality that I need to use in my module, like `rules_python` with its pip setups, I go with a `bazel_dep` (and thus `git_override` if I need to), if I just need it as a build dep (e.g. in a `cc_library` in my case), I go `git_repository`". Does this make sense?

**Reply by fmeum** [reply]
Yes, that's a pretty good rule of thumb!

**Reply by meteorcloudy** [reply]
You might also need to consider if the same dependency could be used by other modules in your dependency graph. 

For example, if multiple modules depending on zlib as a `git_repository` in their own MODULE.bazel files, then they are actually different copies, which could lead to linking conflicts in the final binary. In that case, it'll be better to have zlib as a Bazel module.

**Reply by nicholasjng** [reply]
Thanks! While it's not going to happen in my case, the library I'm trying to modularize (nanobind) is a very natural dependency for Python targets building C++ extensions into their wheels, so it's very good to know about that.

Just out of interest: Is there a "strict mode" for Bazel graph building that disallows copies of `git_repository` (and similar) dependencies for modules? Or do you need to modularize said dependencies in order for that to work?

**Reply by meteorcloudy** [reply]
It's hard for Bazel to enforce this, because Bazel doesn't know all the information and how the dependencies will be used. In some cases, it's totally fine and might be desired to have different copies/versions of the same dependency or to introduce a dependency that only you need with `git_repository` or `http_archive` repo rule.

There is already a nanobind module in BCR: https://registry.bazel.build/modules/nanobind, is that what you are looking for?

**Reply by nicholasjng** [reply]
No, that's not what I'm looking for - nanobind is not a great fit for package managers and registries because it is so configurable and has lots of platform-, Python version-, and feature-dependent optimizations and build flags.

For that reason, nanobind itself recommends building it inplace for your specific application, and I agree with this reasoning, which is why I wanted to find a way to bootstrap it from a git repo using bzlmod. But it seems the `git_repository` rule is the best fit for this need as it stands.

**Reply by meteorcloudy** [reply]
Just to be clear, the current nanobind module is built from source (like most modules in BCR), but I understand it might not be flexible enough to be configured by the end user. However, it's definitely feasible to have better integration with Bzlmod to make it possible.
https://github.com/bazelbuild/bazel-central-registry/blob/main/modules/nanobind/1.2.0/patches/add_build_file.patch#L4

**Reply by nicholasjng** [reply]
Yeah, that's my gripe with that workflow - that tag was pushed for Google benchmark, and I changed some of the build and linker options afterwards.

Also, if you decide to change the options, do you just clobber the tag and change options for everyone in the BCR? I don't think that's good.

**Reply by meteorcloudy** [reply]
No, I meant something like https://docs.google.com/document/d/132vEDnQZY0PtF9ko6HoLQ2dqk6meZnSrUfc1gsKuBkk/edit for https://github.com/bazelbuild/bazel/pull/16552

The end user should have the control. But if having multiple copies for nanobind is generally fine and won't cause any conflict, then yes, probably it's better for everyone to have their own version.

**Reply by nicholasjng** [reply]
Fair, then I still need to pull in the C headers of the target Python version (which is hardcoded to 3.8 in BCR nanobind, oof), but `@rules_python//python/cc:current_py_cc_headers` might let me.

If you provide these symbols for the users, you're giving a lot of value already. But even then, you might still want to give the option of rolling your own module/build file entirely, since the customization level is already considerably high.

By the way, thanks for the comments! It's very valuable to me to have this kind of exchange with actual experts, because (IMO) the docs are not exhaustive in many places for the casual user.
